### PR TITLE
[release 1.1] isolateEmulatorThread: Add full-pcpu-only support

### PR DESCRIPTION
**Pull request by RamLavi**
This is an manual cherry-pick of:
- https://github.com/kubevirt/kubevirt/pull/10593 
- https://github.com/kubevirt/kubevirt/pull/10783.
- https://github.com/kubevirt/kubevirt/pull/10839.
- https://github.com/kubevirt/kubevirt/pull/10872

Together these 4 PRs fix this issue: [CNV-31584](https://issues.redhat.com/browse/CNV-31584), according to this [proposal](https://github.com/kubevirt/community/pull/247).

```release-note
Fixes SMT Alignment Error in virt-launcher pod by optimizing isolateEmulatorThread feature: https://issues.redhat.com/browse/CNV-31584.
```

### CODE REVIEW

**Review by orelmisan — APPROVED** [review]

**Comment by xpivarc** [comment]
/hold
@vladikr Please unhold if you are ok with partial backport.

**Comment by RamLavi** [comment]
@vladikr, the 2nd phase PR https://github.com/kubevirt/kubevirt/pull/10783 is merged, could you please approve the backport and remove the hold?

**Comment by RamLavi** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/6173c93795b206bd41c19bebe9fc446ea365b7b2..d8e6056b48dec35024c30c3424ec8997d52983e5): Rebase

**Comment by RamLavi** [comment]
Change: Cherry-picked the commits from (the now merged) phase 2 PR https://github.com/kubevirt/kubevirt/pull/10783 in order to backport the full bug-fix in one PR.
@vladikr could you take a look?

**Comment by RamLavi** [comment]
/hold
On second thought, let's wait for the discussion with our stakeholders before we persue with the backport

**Comment by RamLavi** [comment]
rearranging the commits needed for this backport

**Review by orelmisan — APPROVED** [review]

**Comment by RamLavi** [comment]
/hold cancel
removing hold as there is an agreement on implementation, and all main branch PRs are merged.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go
Comment (xpivarc) [inline pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go:100]: Missing dedicated cpu check.

**Comment by xpivarc** [comment]
I would recommend squashing the commits but I will understand if you prefer not to do it.

File: pkg/virt-controller/services/renderresources.go
Comment (RamLavi) [reply pkg/virt-controller/services/renderresources.go:219]: it is confusing. I will want to refactor this in a separate PR (without changing the logic of course)

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go]: We use it to create the emulatorThreadPin, and then pass it to the ioThread formatting function

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go
Comment (RamLavi) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go:100]: We discussed it in the original PR, there is currently a validating webhook preventing the user from setting `IsolateEmulatorThread` without `dedicatedCPUs`. This is why in the code only IsolateEmulatorThread is checked.
However, I think that if we want to fix it we need to do it on a separate PR.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:419]: I see. this is not the behavior used. IIUC the log here is used to increase the verbosity of the error.
However I can change it if you want - but can I set a change on a separate PR to align all the places at once?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go
Comment (RamLavi) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: this is later overridden.

**Comment by RamLavi** [comment]
@xpivarc, considering that the PRs were approved on main branch, and if you don't have any big objections to the code, would it be OK if you approve it and we will issue a separate PR(s) with the fixes?

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go
Comment (xpivarc) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go:100]: You are right, thanks for the reminder.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go
Comment (xpivarc) [inline pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:485]: Hmm, I am not sure this will work with hotplug. Do we need the vCPUs?

**Comment by xpivarc** [comment]
> @xpivarc, considering that the PRs were approved on main branch, and if you don't have any big objections to the code, would it be OK if you approve it and we will issue a separate PR(s) with the fixes?

I am okay with making an exception as this is a backport. I will be more than happy if we can clean the main.

I left a second review where I focused if this was the correct change. I think I found one thing that might be problematic.

File: pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go
Comment (RamLavi) [reply pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:485]: I think we addressed the CPU hotplug corner [here](https://github.com/kubevirt/kubevirt/pull/10757/files#diff-98b8339212e612cd92ce33649e492336892a3ed649f368b181aad9aa2854138cR484). I consulted with @enp0s3 and also checked it manually at the time (the original PR).

**Review by xpivarc — COMMENTED** [review]
/approve

**Pull request merged**
